File: 10단계/10870.java
import java.util.*;
import java.io.*;

public class Main{
    static BufferedReader br;
    static BufferedWriter bw;
    static int[] cache;
    public static void main(String[] args) throws IOException{
        br = new BufferedReader(new InputStreamReader(System.in));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(br.readLine());
        cache = new int[n+1];
        Arrays.fill(cache,-1);
        bw.write(String.valueOf(recur(n)));
        bw.flush();bw.close();br.close();
    }
    public static int recur(int idx){
        if(idx == 0 || idx == 1){
            return idx;
        }
        if(cache[idx] != -1)
            return cache[idx];
        int tmp = recur(idx-1) + recur(idx-2);
        cache[idx] = tmp;
        return tmp;
    }
}

File: Archived/재귀/10870.py
import sys
sys.setrecursionlimit(10000000)

def recur(n):
    global cache
    if n == 0 or n == 1:
        cache[n] = n
        return n
    if cache[n] != None:
        return cache[n]
    
    tmp = recur(n-1) + recur(n-2)
    cache[n] = tmp
    return tmp

n = int(sys.stdin.readline().strip())
cache = [None for i in range(n+1)]
print(recur(n))
